### PR TITLE
Fix performance regression in debuginfo file_metadata.

### LINKED ISSUE #70785: A big performance regression in benchmarks
[rustc performance data](https://perf.rust-lang.org/compare.html?start=74bd074eefcf4915c73d1ab91bc90859664729e6&end=6050e523bae6de61de4e060facc43dc512adaccd&stat=instructions:u) shows a pretty big regression between [74bd074e](https://github.com/rust-lang/rust/commit/74bd074eefcf4915c73d1ab91bc90859664729e6) and [6050e52](https://github.com/rust-lang/rust/commit/6050e523bae6de61de4e060facc43dc512adaccd). I don't know if this is expected, but I'm putting this out there in case others see the dent in perf charts and wonder what had happened.

<hr/>

**EDIT**(@eddyb): we how have a [single-PR perf comparison showing the regression](https://perf.rust-lang.org/compare.html?start=9e55101bb681010c82c3c827305e2665fc8f2aa0&end=6050e523bae6de61de4e060facc43dc512adaccd&stat=instructions:u) for #69718.

#70156, the other PR in the range, [is much more of a mixed bag](https://perf.rust-lang.org/compare.html?start=74bd074eefcf4915c73d1ab91bc90859664729e6&end=9e55101bb681010c82c3c827305e2665fc8f2aa0&stat=instructions:u) (but only for a small subset of runs).

### PR DESCRIPTION
**Pull request by arlosi**
Fixes performance regression caused by #69718.

Finding the `SourceFile` associated with a `FileName` called `get_source_file` on the `SourceMap`, which does a linear search through all files in the `SourceMap`.

This resolves the issue by passing the `SourceFile` in from the caller (which already had it available) instead of the `FileName`

Fixes #70785.

### CODE REVIEW

**Comment by arlosi** [comment]
r? @eddyb 

Should we do a perf run to see if this fully resolves the regression? It seemed to resolve it in a few benchmarks I ran locally.

File: src/librustc_codegen_llvm/debuginfo/metadata.rs [hunk 1]
```diff
@@ -761,14 +761,13 @@ fn hex_encode(data: &[u8]) -> String {
...
 pub fn file_metadata(
     cx: &CodegenCx<'ll, '_>,
-    file_name: &FileName,
+    source_file: &SourceFile,
     defining_crate: CrateNum,
 ) -> &'ll DIFile {
-    debug!("file_metadata: file_name: {}, defining_crate: {}", file_name, defining_crate);
+    debug!("file_metadata: file_name: {}, defining_crate: {}", source_file.name, defining_crate);
 
-    let source_file = cx.sess().source_map().get_source_file(file_name);
```
Comment (eddyb) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs:769]: Ahh, I missed this during review! This is indeed unnecessarily slow.

File: src/librustc_codegen_llvm/debuginfo/metadata.rs [hunk 2]
```diff
@@ -761,14 +761,13 @@ fn hex_encode(data: &[u8]) -> String {
 
 pub fn file_metadata(
     cx: &CodegenCx<'ll, '_>,
-    file_name: &FileName,
+    source_file: &SourceFile,
```
Comment (eddyb) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs:764]: Heh, I think I did this (plus removing `defining_crate` and relying on the more accurate `source_file.is_imported()`) in one of my PRs that has been sitting around.

**Comment by eddyb** [comment]
Let's land it and see the results there, to reduce the time `master` spends regressed.

@bors r+ rollup=never p=10

**Comment by nnethercote** [comment]
@arlosi: This change fixes most of the regressions, thanks.

But I notice that the "clap-rs-debug patched incremental: println" benchmark regression is only partly fixed. Any ideas what happened there?

**Comment by mati865** [comment]
@nnethercote this PR has fixed the regression: https://perf.rust-lang.org/compare.html?start=9e55101bb681010c82c3c827305e2665fc8f2aa0&end=607b8582362be8e26df7acc12fa242359d7edf95&stat=instructions:u
Clap regressed right before original PR due to https://github.com/rust-lang/rust/pull/70156#issuecomment-602486422

**Comment by nnethercote** [comment]
@mati865: is that the right perf link? Those perf results look like not much changed.

**Comment by mati865** [comment]
@nnethercote it's difference starting right before original PR and after this fix proving there is no big preformance hit.
For clap see this results: https://perf.rust-lang.org/compare.html?start=74bd074eefcf4915c73d1ab91bc90859664729e6&end=9e55101bb681010c82c3c827305e2665fc8f2aa0&stat=instructions:u and the comment I linked above for the explanation.

**Comment by nnethercote** [comment]
I see now that this caused a large regression in [memory usage](https://perf.rust-lang.org/compare.html?start=7b657d340d715f48449189fea9d032350323a13f&end=607b8582362be8e26df7acc12fa242359d7edf95&stat=max-rss)

@arlosi: was that expected? Any thoughts on how it could be improved?

**Comment by arlosi** [comment]
@nnethercote  It appears that the change that caused the CPU regression, also caused an rss improvement. And that fixing the CPU regression removed the rss improvement.

The original change that contained the CPU regression went in on 2020-04-03, and the fix went in on 2020-04-05. Looking at the graph for cargo-debug max-rss over that time period shows an improvement in memory usage, then a matching regression:
https://perf.rust-lang.org/index.html?start=2020-04-01&end=2020-04-08&absolute=true&stat=max-rss

Here's the run that introduced the CPU regression (and rss improvement):
https://perf.rust-lang.org/compare.html?start=9e55101bb681010c82c3c827305e2665fc8f2aa0&end=6050e523bae6de61de4e060facc43dc512adaccd&stat=max-rss

The key difference that caused the CPU regression was adding this line:
    `let source_file = cx.sess().source_map().get_source_file(file_name);`
However, I can't explain how that could have reduced memory usage.

**Comment by nnethercote** [comment]
I investigated the max RSS change. As far as I can tell, the number of allocations, and their sizes, are basically identical in the two versions. But the distribution is different, due to the significant timing differences. My theory is that because the back-end is multi-threaded, how the memory peaks for all the individual threads overlap have a big effect on how the global peak manifests.

Here is what DHAT tells me about memory usage when the speed regression was in place:
```
    Total:     22,161,188,077 bytes (100%, 71,870.02/Minstr) in 52,214,866 blocks (100%, 169.34/Minstr), avg size 424.42 bytes, avg lifetime 6,620,711,098.71 instrs (2.15% of program duration)
    At t-gmax: 662,297,094 bytes (100%) in 2,100,503 blocks (100%), avg size 315.3 bytes
```
and here's what it looks like now:
```
    Total:     22,161,559,231 bytes (100%, 128,445.93/Minstr) in 52,214,794 blocks (100%, 302.63/Minstr), avg size 424.43 bytes, avg lifetime 7,079,327,664.2 instrs (4.1% of program duration)
    At t-gmax: 1,018,333,644 bytes (100%) in 4,138,546 blocks (100%), avg size 246.06 bytes
```
The "Total" lines describe the allocations done over the entire life of the program, and they are the same. The "At t-gmax" lines describe the live allocations that exist at the global memory peak, and they are very different.

Here is the with-regression memory usage shown by Massif:
old
There are lots of local maxima, spread out evently, that are close to the global maximum.

And here is the same thing for the current code:
new
The global maximum is larger and more pronounced relative to the rest of the curve.

In short, I don't think there's really anything to be done about this. At least it's good to know that nothing silly is happening.
